<div align="center">

# Modelo de Clasificación del Riesgo de Entrega Tardía en DataCo

### Big Data

**Jonathan Barrantes · Joy D. Chávez Pomacaja · María de Jesús Peña García · Abby Chavarría Bolaños**

Universidad Fidélitas  
San José, Costa Rica

</div>

---

### Autores y contacto

| Integrante                 | Correo institucional        |
| -------------------------- | --------------------------- |
| Jonathan Barrantes         | jbarrantes40180@ufide.ac.cr |
| Joy D. Chávez Pomacaja     | jchavez60736@ufide.ac.cr    |
| María de Jesús Peña García | mpena30226@ufide.ac.cr      |
| Abby Chavarría Bolaños     | achavarria70596@ufide.ac.cr |

---

## Resumen

> Este proyecto busca predecir el riesgo de entregas tardías en DataCo mediante un modelo de clasificación basado en aprendizaje automático. Se utilizará el conjunto de datos **DataCo Smart Supply Chain** y la variable objetivo `Late_delivery_risk`.
>
> Se explorarán y prepararán los datos, se analizarán las características asociadas con los retrasos y se comparará el desempeño de distintos modelos. Las predicciones utilizarán información disponible al registrar el pedido, excluyendo datos posteriores que revelen el resultado de la entrega.
>
> Se pretende generar estimaciones que permitan identificar pedidos con mayor riesgo y apoyar la planificación logística.

**Palabras clave:** aprendizaje automático · cadena de suministro · clasificación · entregas tardías · planificación logística

---

## I. ANTECEDENTES

DataCo Global inició sus operaciones en 2012, dedicada a la comercialización y distribución de productos en el mercado nacional. En sus primeros años, el volumen reducido de pedidos permitía organizar las entregas mediante plazos generales y seguimiento individual.

A partir de 2015, su expansión hacia mercados internacionales incrementó los pedidos, los destinos y las modalidades de envío.

Con este crecimiento, las prácticas iniciales de planificación resultaron insuficientes y comenzaron a presentarse retrasos recurrentes. La empresa atendía los incumplimientos principalmente después de recibir reclamos, lo que dificultaba actuar anticipadamente.

Al mismo tiempo, la competencia y las expectativas de los clientes aumentaron la presión por cumplir los plazos de entrega.

---

## II. DEFINICIÓN DEL PROBLEMA

DataCo presenta limitaciones para anticipar las entregas tardías tras la expansión de sus operaciones. Su planificación continúa utilizando plazos generales que no consideran suficientemente las diferencias entre destinos, modalidades de envío y características de los pedidos.

Además, el seguimiento se concentra en atender incumplimientos después de recibir reclamos, lo que dificulta identificar oportunamente los pedidos con mayor riesgo y priorizar su atención.

Estas limitaciones afectan la planificación de acciones preventivas y pueden ocasionar mayores costos, acumulación de reclamos e insatisfacción de los clientes.

El problema se centra, por tanto, en la dificultad de anticipar posibles retrasos y orientar el seguimiento de los pedidos antes de que se incumplan los plazos.

---

## III. JUSTIFICACIÓN

El proyecto permitiría generar estimaciones del riesgo de entrega tardía para apoyar la planificación logística de DataCo. Esta información facilitaría la identificación de pedidos que requieren mayor atención y la priorización de su seguimiento antes del vencimiento de los plazos.

Asimismo, el análisis de las características de los pedidos, sus destinos y modalidades de envío aportaría criterios para orientar decisiones preventivas.

Estos beneficios podrían contribuir a mejorar la puntualidad de las entregas, controlar los costos asociados con los retrasos y favorecer la satisfacción de los clientes.

---

## IV. OBJETIVO GENERAL

> Desarrollar un modelo de clasificación mediante técnicas de Big Data y aprendizaje automático para predecir el riesgo de entregas tardías en DataCo y apoyar la planificación logística.

---

## V. OBJETIVOS ESPECÍFICOS

1. Analizar las características geográficas, logísticas y de los pedidos presentes en el conjunto de datos, identificando sus relaciones con el riesgo de entrega tardía.

2. Evaluar la calidad de los datos mediante la identificación de valores faltantes, inconsistencias, valores atípicos y distribuciones que puedan afectar el desempeño de los modelos.

3. Preparar los datos mediante técnicas de limpieza, transformación, codificación, escalamiento e imputación, según las necesidades identificadas.

4. Seleccionar y construir variables predictoras relevantes para estimar el riesgo de entrega tardía, considerando únicamente la información disponible al registrar el pedido.

5. Construir distintos modelos de clasificación mediante algoritmos de aprendizaje supervisado para estimar el riesgo de entrega tardía de los pedidos.

6. Comparar el desempeño de los modelos mediante procedimientos de validación y métricas de clasificación, como exactitud, precisión, exhaustividad y F1-score, para determinar su capacidad de identificar pedidos con riesgo de retraso.

7. Ajustar y seleccionar el modelo que presente un equilibrio adecuado entre capacidad predictiva y complejidad, y evaluar su desempeño final sobre un conjunto de prueba reservado para estimar su capacidad de generalización.

---

## VI. DESCRIPCIÓN DEL CONJUNTO DE DATOS

El proyecto utilizará el archivo `DataCoSupplyChainDataset.csv` del conjunto **DataCo Smart Supply Chain**, disponible públicamente en Kaggle [1].

Este contiene información histórica sobre clientes, productos, pedidos y operaciones de envío de DataCo Global.

| Característica                        | Descripción                                                                                                                                                                                            |
| ------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **Estructura**                        | Contiene **180 519 registros y 53 variables**. Incluye identificadores de pedido y de artículo, por lo que el número de registros no debe interpretarse directamente como el número de pedidos únicos. |
| **Variable objetivo**                 | `Late_delivery_risk` es una variable binaria que identifica registros etiquetados como entrega tardía mediante `1` y registros no etiquetados como tardíos mediante `0`.                               |
| **Variables categóricas**             | Incluye modalidad de envío, categoría del producto, mercado, región de destino, tipo de transacción y segmento del cliente.                                                                            |
| **Variables numéricas**               | Incluye días programados de envío, cantidad de artículos, precios, descuentos, importes de venta y ganancias.                                                                                          |
| **Variables temporales**              | Contiene fechas de pedido y envío. La fecha de registro del pedido permitirá obtener características como mes y día de la semana.                                                                      |
| **Información posterior a la compra** | Incluye días reales de envío y estados de entrega. Estos campos podrán utilizarse para el análisis histórico y la revisión de las etiquetas, pero se excluirán de las entradas del modelo predictivo.  |

### Variable objetivo

| Valor | Clasificación       |
| :---: | ------------------- |
|  `0`  | Entrega sin retraso |
|  `1`  | Entrega tardía      |

---

## VII. REFERENCIAS

**[1]** S. Tiwari, _DataCo SMART SUPPLY CHAIN FOR BIG DATA ANALYSIS_, Kaggle, conjunto de datos. [En línea]. Disponible en:  
https://www.kaggle.com/datasets/shashwatwork/dataco-smart-supply-chain-for-big-data-analysis  
[Consultado: 22 de septiembre de 2026].


# Análisis exploratorio y preparación inicial de los datos

## 1. Carga y exploración inicial del conjunto de datos


In [12]:
### 1.1 Importación de librerías
import pandas as pd
import numpy as np

In [13]:
### 1.2 Carga del conjunto de datos
df = pd.read_csv(
    "Dataset/DataCoSupplyChainDataset.csv",
    encoding="latin1",
    sep=";"
)

df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Name,Customer City,Customer Country,...,Order State,Order Status,Order Zipcode,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,Sporting Goods,Caguas,Puerto Rico,...,Java Occidental,COMPLETE,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,Sporting Goods,Caguas,Puerto Rico,...,Rajastï¿,PENDING,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,Sporting Goods,San Jose,EE. UU.,...,Rajastï¿,CLOSED,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,Sporting Goods,Los Angeles,EE. UU.,...,Queensland,COMPLETE,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,Sporting Goods,Caguas,Puerto Rico,...,Queensland,PENDING_PAYMENT,NaN,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [14]:
## 1.3 Verificacion de Estructura del CSV
print(f"Registros: {df.shape[0]:,}")
print(f"Variables: {df.shape[1]}")

Registros: 180,519
Variables: 44


In [15]:
### 1.4 Estructura y tipos de datos
resumen_estructura = pd.DataFrame({
    "Variable": df.columns,
    "Tipo de dato": df.dtypes.astype(str).values
})

print(f"Total de variables: {df.shape[1]}")
display(resumen_estructura)

Total de variables: 44


,Variable,Tipo de dato
0,Type,str
1,Days for shipping (real),int64
2,Days for shipment (scheduled),int64
3,Benefit per order,float64
4,Sales per customer,float64
5,Delivery Status,str
6,Late_delivery_risk,int64
7,Category Name,str
8,Customer City,str
9,Customer Country,str


In [16]:
### 1.5 Valores faltantes
resumen_nulos = pd.DataFrame({
    "Valores faltantes": df.isnull().sum(),
    "Porcentaje (%)": (df.isnull().sum() / len(df) * 100).round(2)
})

resumen_nulos = resumen_nulos[resumen_nulos["Valores faltantes"] > 0]
resumen_nulos = resumen_nulos.sort_values(
    by="Valores faltantes",
    ascending=False
)

display(resumen_nulos)

,Valores faltantes,Porcentaje (%)
Product Description,180519,100.00
Order Zipcode,155679,86.24
Customer Lname,8,0.00
Customer Zipcode,3,0.00


Se identificaron valores faltantes en cuatro variables. `Product Description` presenta un **100 % de valores faltantes**, mientras que `Order Zipcode` presenta un **86.24 %**. Por su parte, `Customer Lname` y `Customer Zipcode` presentan únicamente 8 y 3 valores faltantes, respectivamente, por lo que su proporción respecto al total de registros es mínima.

En esta etapa únicamente se documentan los valores faltantes identificados. Su tratamiento se determinará posteriormente durante la preparación de los datos.


In [17]:
### 1.6 Estadísticas descriptivas
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Days for shipping (real),180519.0,3.497654,1.623722,0.000000,2.000000,3.000000,5.000000,6.000000
Days for shipment (scheduled),180519.0,2.931847,1.374449,0.000000,2.000000,4.000000,4.000000,4.000000
Benefit per order,180519.0,21.974989,104.433526,-4274.979980,7.000000,31.520000,64.800003,911.799988
Sales per customer,180519.0,183.107609,120.043670,7.490000,104.379997,163.990005,247.399994,1939.989990
Late_delivery_risk,180519.0,0.548291,0.497664,0.000000,0.000000,1.000000,1.000000,1.000000
Customer Zipcode,180516.0,35921.126914,37542.461122,603.000000,725.000000,19380.000000,78207.000000,99205.000000
Latitude,180519.0,29.719955,9.813646,-33.937553,18.265432,33.144863,39.279617,48.781933
Longitude,180519.0,-84.915675,21.433241,-158.025986,-98.446312,-76.847908,-66.370583,115.263077
Order Item Discount,180519.0,20.664741,21.800901,0.000000,5.400000,14.000000,29.990000,500.000000
Order Item Discount Rate,180519.0,0.101668,0.070415,0.000000,0.040000,0.100000,0.160000,0.250000


Las estadísticas descriptivas muestran diferencias importantes en los rangos y escalas de las variables numéricas. Las variables monetarias presentan una mayor amplitud de valores, mientras que otras variables, como tasas, cantidades y días de envío, se encuentran en rangos considerablemente menores.

También se identifican variables numéricas que no deben interpretarse como continuas, como `Late_delivery_risk`, los códigos postales y `Product Status`. Además, `Product Description` no contiene observaciones válidas y `Product Status` presenta un único valor en todos los registros.

Algunas variables presentan estadísticas descriptivas idénticas, como `Benefit per order` y `Order Profit Per Order`, así como `Sales per customer` y `Order Item Total`. Estas posibles redundancias deberán revisarse posteriormente durante la preparación de los datos.


In [18]:
### 1.7 Identificación de variables continuas
variables_continuas = [
    "Benefit per order",
    "Sales per customer",
    "Latitude",
    "Longitude",
    "Order Item Discount",
    "Order Item Discount Rate",
    "Order Item Product Price",
    "Order Item Profit Ratio",
    "Sales",
    "Order Item Total",
    "Order Profit Per Order",
    "Product Price"
]

print(f"Variables continuas identificadas: {len(variables_continuas)}")
variables_continuas

Variables continuas identificadas: 12


['Benefit per order',
 'Sales per customer',
 'Latitude',
 'Longitude',
 'Order Item Discount',
 'Order Item Discount Rate',
 'Order Item Product Price',
 'Order Item Profit Ratio',
 'Sales',
 'Order Item Total',
 'Order Profit Per Order',
 'Product Price']

## 2. Análisis de distribuciones

**Asignado: Joy D. Chávez Pomacaja**

Generar los histogramas de las variables continuas e identificar cuáles presentan distribuciones simétricas, sesgadas, multimodales, acotadas o uniformes.

## 3. Análisis de escalas y posibles tratamientos

**Asignado: María de Jesús Peña García**

Comparar las escalas de las variables, documentar posibles tratamientos futuros e identificar una o varias variables que podrían utilizarse para la estratificación.

## 4. Conjuntos de entrenamiento y prueba

**Asignado: Abby Chavarría Bolaños**

Crear las divisiones aleatoria y estratificada del conjunto de datos y comparar ambas mediante sus proporciones y el error de muestreo.
